Control Plane Agent POC with dapr (#57)
kyma-project/control-plane: docs/internal/investigations/runtime-governor-poc.md

```

To prove that the Runtime configuration can be dynamically reloaded, change the configuration for Runtime 2 to point to the Redis 1 instance. To do so, run the following command:

```bash
CP_IPADDR=$(docker inspect k3d-governor-server-0 --format='{{json .NetworkSettings.Networks.cpnet.IPAddress}}'); REDIS1_PASSWORD=$(kubectl --context=k3d-governor get secret -n redis1-system redis -ojsonpath='{.data.redis-password}' | base64 -d) helm --kube-context=k3d-governor upgrade -i cp-governor -n cp-poc ./governor/chart \
  --set redis1.host=${CP_IPADDR//\"}:30000 --set redis1.password=$REDIS1_PASSWORD \
  --set redis2.host=${CP_IPADDR//\"}:30000 --set redis2.password=$REDIS1_PASSWORD
```

Now you can watch how the configuration is reloaded by the Runtime Agent on Runtime 2. The NodeApp Pod should be restarted.

```bash
watch kubectl --context=k3d-runtime2 get po --all-namespaces
```

### Cleanup

To clean up the POC environment, use the following command:

```bash
./cleanup.sh
```

## Reloading configuration of Dapr sidecars 

> We need to configure Dapr sidecars in a Kyma Runtime from one central Kyma Control Plane. However, he Dapr sidecars have an [issue](https://github.com/dapr/dapr/issues/1172) concerning the fact that the whole Pod needs to restart for the configuration to be applied.

Reviewer: ```suggestion
We need to configure Dapr sidecars in a Kyma Runtime from one central Kyma Control Plane. However, Dapr sidecars have an [issue](https://github.com/dapr/dapr/issues/1172) concerning the fact that the whole Pod needs to restart for the configuration to be applied.
```

> There are a few solutions for this problem:

Reviewer: ```suggestion
There are a few solutions to this problem.
```

I would avoid using a colon `:` to introduce headings.

> ### Control Plane Runtime Agent will restart every Pod with a Dapr sidecar

Reviewer: ```suggestion
### Control Plane Runtime Agent restarts every Pod with a Dapr sidecar
```

This solution may be the simplest, but it could add too many responsibilities for one component. A single Control Plane Runtime Agent would fetch the configuration, apply it, and then restart all Pods.

Pros:
- No new components in the Runtime 

> - Less resource-consuming

Reviewer: ```suggestion
- Less resource consumption
```
or `less resource-consuming <something>`

Cons:
- Too many responsibilites for one component

### New component responsible for restarting Pods

> This solution adds a new component that will be triggered by the Agent and will restart Pods mentioned in the configuration.

Reviewer: ```suggestion
This solution adds a new component that will be triggered by the Runtime Agent and will restart Pods mentioned in the configuration.
```

Pros:
- Responsibilites distributed between components

Cons:

> - Another component will definitely mean more resource consumption, which we want to avoid

Reviewer: ```suggestion
- Another component would definitely mean more resource consumption, which we want to avoid
```

> ### Dapr contribution that would allow configuration reloading

Reviewer: ```suggestion
### Dapr contribution allowing configuration reloading
```

> This solution will surely be the most time-consuming. It seems like a really big change to the Dapr ecosystem and the approach to that is still being [discussed](https://github.com/dapr/dapr/issues/1172#issuecomment-610568718). We would have to create a proposal, wait for it to be accepted, and then develop the feature.

Reviewer: ```suggestion
This solution would surely be the most time-consuming. It seems like a really big change to the Dapr ecosystem and the approach to that [is still being discussed](https://github.com/dapr/dapr/issues/1172#issuecomment-610568718). We would have to create a proposal, wait for it to be accepted, and then develop the feature.
```

Pros:

> - Configuration reload out-of-the-box

Reviewer: ```suggestion
- Out-of-the-box configuration reload 
```
with hyphens when it's an adjective
or 
`Configuration reload out of the box`
with no hyphens when it's an adverbial phrase of manner.
